platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,493 @@
+An exception will be thrown if the Roster is invalid.
+The durability of the submitted Roster is ensured via storing it in the state.
+
+The existing `PlatformState` class is in the middle of a refactor, and in the long term, may cease to exist.
+However, its replacement will continue to have this new API to set a candidate roster.
+
+## Data Structure
+
+The App is already responsible for managing the address book.
+We propose that it continues to do so, and when it receives a HAPI transaction, creates a candidate Roster object
+and passes it to the platform via the new API. The platform will then validate the roster and store it in the state.
+
+### Protobuf
+
+#### Roster
+
+The protobuf definition for the Roster will look as follows:
+
+```proto
+/**
+ * A single roster in the network state.
+ * <p>
+ * The roster SHALL be a map of key `NodeId`, value `RosterEntry`.
+ */
+message Roster {
+
+    /**
+     * Map of roster entries, one per consensus node.
+     * <p>
+     * This map SHALL contain roster entries in natural order of ascending node ids.
+     * The node Id key SHALL match the node Id in the value.
+     * This map SHALL NOT be empty.<br/>
+     */
+    map<uint64, RosterEntry> rosterEntries = 1;
+}
+```
+
+#### RosterEntry
+
+A roster entry will look as follows:
+
+```proto
+/**
+ * A single roster entry in the network state.
+ *
+ * Each roster entry in SHALL encapsulate the elements required
+ * to manage node participation in the Threshold Signature Scheme (TSS).<br/>
+ * All fields are REQUIRED.
+ */
+message RosterEntry {
+
+    /**
+     * A consensus node identifier.
+     * <p>
+     * Node identifiers SHALL be unique _within_ a shard and realm,
+     * but a node SHALL NOT, ever, serve multiple shards or realms,
+     * therefore the node identifier MAY be repeated _between_ shards and realms.
+     */
+    uint64 node_id = 1;
+
+    /**
+     * A consensus weight.
+     * <p>
+     * Each node SHALL have a weight of zero or more in consensus calculations.<br/>
+     * The consensus weight of a node SHALL be calculated based on the amount
+     * of HBAR staked to that node.<br/>
+     * Consensus SHALL be calculated based on agreement of greater than `2/3`
+     * of the total `weight` value of all nodes on the network.
+     */
+    uint64 weight = 2;
+
+    /**
+     * An RSA public certificate used for signing gossip events.
+     * <p>
+     * This value SHALL be a certificate of a type permitted for gossip
+     * signatures.<br/>
+     * This value SHALL be the DER encoding of the certificate presented.<br/>
+     * This field is REQUIRED and MUST NOT be empty.
+     */
+    bytes gossip_ca_certificate = 3;
+
+    /**
+     * An elliptic curve public signing key.<br/>
+     * This contains the _long term_ public key for this node.
+     * <p>
+     * This value SHALL be the DER encoding of the presented elliptic curve
+     * public key.<br/>
```

Comment:
Should we spec which kinds of key curves are supported?

Reply:
I'm not sure. This is a byte field, so the supported key curves will not change its invariant if/when they change. We may opt to document this fact for informational purposes, but that should live in the java code.

Reply:
I think we should indicate the curve, just in case someone doesn't trust our CLI tool and wants to generate their `tssEncryptionKey` with another tool.

Reply:
Why is there not an RSA encryption key field? The roster used by the consensus system needs a key field and curve type so the event intake validation can verify the event, but doesn't specifically need to know anything about TSS. On the other hand, the roster used by the execution system also needs to save in state the TSS key (I assume), in which case we need both types of fields here?